kong/clustering/data_plane.lua

```diff
@@ -175,55 +170,44 @@ function _M:communicate(premature)
   --                 and is also responsible for handling timeout detection
 
   local ping_immediately
-  local config_exit
   local next_data
 
-  local config_thread = ngx.thread.spawn(function()
-    while not exiting() and not config_exit do
-      local ok, err = config_semaphore:wait(1)
-      if ok then
-        local data = next_data
-        if data then
-          local msg = assert(inflate_gzip(data))
-          yield()
-          msg = assert(cjson_decode(msg))
-          yield()
-
-          if msg.type == "reconfigure" then
-            if msg.timestamp then
-              ngx_log(ngx_DEBUG, _log_prefix, "received reconfigure frame from control plane with timestamp: ",
-                                 msg.timestamp, log_suffix)
-
-            else
-              ngx_log(ngx_DEBUG, _log_prefix, "received reconfigure frame from control plane", log_suffix)
-            end
-
-            local config_table = assert(msg.config_table)
-            local pok, res
-            pok, res, err = pcall(config_helper.update, self.declarative_config,
-                                  config_table, msg.config_hash, msg.hashes)
-            if pok then
-              if not res then
-                ngx_log(ngx_ERR, _log_prefix, "unable to update running config: ", err)
-              end
-
-              ping_immediately = true
-
-            else
-              ngx_log(ngx_ERR, _log_prefix, "unable to update running config: ", res)
-            end
-
-            if next_data == data then
-              next_data = nil
-            end
-          end
-        end
+  local handler = function(data)
+    if exiting() or not data then
+      return
+    end
 
-      elseif err ~= "timeout" then
-        ngx_log(ngx_ERR, _log_prefix, "semaphore wait error: ", err)
-      end
+    local msg = assert(inflate_gzip(data))
+    yield()
+    msg = assert(cjson_decode(msg))
+    yield()
+
+    if msg.type ~= "reconfigure" then
+      return
     end
-  end)
+
+    ngx_log(ngx_DEBUG, _log_prefix, "received reconfigure frame from control plane",
+                       msg.timestamp and " with timestamp: " .. msg.timestamp or "",
+                       log_suffix)
+
+    local config_table = assert(msg.config_table)
+
+    local pok, res, err = pcall(config_helper.update, self.declarative_config,
+                                config_table, msg.config_hash, msg.hashes)
+    if not pok or not res then
+      ngx_log(ngx_ERR, _log_prefix, "unable to update running config: ", err)
+
+    else
+      ping_immediately = true
```

Comment:
If I understood this correctly this logic seemed to have been changed. Now it is `pok and res` which means in case of config load failure, we no longer send a ping as previously.

Reply:
Good catch, thank you, I will revert the code.